reps/2024-10-18-train-tune-api-revamp/2024-10-18-train-tune-api-revamp.md

```diff
@@ -0,0 +1,1194 @@
+3. UX friction:
+   * There is significant overlap in functionality with the `resume_from_checkpoint` API, and distinguishing the two is a common user question \[[1](https://ray.slack.com/archives/CSX7HVB5L/p1707241076911069), [2](https://ray.slack.com/archives/CSX7HVB5L/p1715627477633579), [3](https://github.com/ray-project/ray/issues/40626)\].
+   * Users need to concatenate the `storage_path` and `name` together to pass into the `restore` API, rather than keeping them separate like they are in `RunConfig`. This is especially unwieldy when cloud storage is used, and users need to resort to manual string joins.
+
+The new design avoids pickling any configurations, and uses the `RunConfig(storage_path, name)` as the unique identifier for a restorable run. If a previous run has written outputs and checkpoints to `{storage_path}/{name}`, then a new run started with the same configuration will “restore” from the state found at that directory. At the moment, the only stateful component is Ray Train’s checkpoint manager, which tracks the latest checkpoint.
+
+Here’s an example of what job-level restoration looks like before and after, where the job is submitted with this `python main.py <UNIQUE_JOB_ID>` as the entrypoint:
+
+<table>
+<!-- Table Headers -->
+  <tr>
+    <th>Before</th>
+    <th>After</th>
+  </tr>
+  <tr>
+<td>
+
+```python
+# main.py
+
+def train_fn_per_worker(config):
+    checkpoint = ray.train.get_checkpoint()
+
+    ...
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("run_id", type=str)
+    args = parser.parse_args()
+
+    unique_run_name = args.run_id  # some UNIQUE job id
+    storage_path = "s3://my/bucket"
+
+    run_path = f"{storage_path}/{unique_run_name}"
+
+    if TorchTrainer.can_restore(run_path):
+        # Certain configs get loaded from a pkl file,
+        # and other parameters need to be re-specified.
+        trainer = TorchTrainer.restore(
+            run_path,
+            train_loop_per_worker=train_fn_per_worker,
+            datasets={...},
+        )
+    else:
+
+        trainer = TorchTrainer(
+            train_fn_per_worker,
+            ...,
+            run_config=ray.train.RunConfig(
+                storage_path=storage_path,
+                name=unique_run_name,
+            ),
+        )
+
+    trainer.fit()
+```
+
+</td>
+<td>
+
+```python
+# main.py
+
+def train_fn_per_worker(config):
+    checkpoint = ray.train.get_checkpoint()
+
+    ...
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("run_id", type=str)
+    args = parser.parse_args()
+
+    unique_run_name = args.run_id  # some UNIQUE job id
+    storage_path = "s3://my/bucket"
+
+    trainer = TorchTrainer(
+        train_fn_per_worker,
+        ...,
+        run_config=ray.train.RunConfig(
```

Comment:
A few questions

1. If there is nothing to restore, e.g., this is the first time the job is submitted, then `unique_run_name` will be `None` and the `TorchTrainer` will see that there is no checkpoint to load from and will not load anything?
2. What's the reason for removing the explicit `restore` call?
3. Why do I need the storage info in `RunConfig`? Couldn't that all just go into `get_checkpoint`?